kevm-pyk/src/kontrol/foundry.py

```diff
@@ -758,37 +800,87 @@ def _init_and_run_proof(_init_problem: tuple[str, str, str | None]) -> tuple[boo
-            return passed, failure_log
-
-    def run_cfg_group(
-        tests: list[tuple[str, str | None]]
-    ) -> dict[tuple[str, str | None], tuple[bool, list[str] | None]]:
-        def _split_test(test: tuple[str, str | None]) -> tuple[str, str, str | None]:
-            test_name, id = test
-            contract, method = test_name.split('.')
-            return contract, method, id
-
-        init_problems = [_split_test(test) for test in tests]
-
-        _apr_proofs: list[tuple[bool, list[str] | None]]
-        if workers > 1:
-            with ProcessPool(ncpus=workers) as process_pool:
-                _apr_proofs = process_pool.map(_init_and_run_proof, init_problems)
-        else:
-            _apr_proofs = []
-            for init_problem in init_problems:
-                _apr_proofs.append(_init_and_run_proof(init_problem))
+            return proof
+
+    def run_cfg_group(tests: list[tuple[str, str | None]]) -> list[APRProof]:
+        llvm_definition_dir = foundry.llvm_library if use_booster else None
+
+        def create_server() -> KoreServer:
+            return kore_server(
+                definition_dir=foundry.kevm.definition_dir,
+                llvm_definition_dir=llvm_definition_dir,
+                module_name=foundry.kevm.main_module,
+                command=kore_rpc_command,
+                bug_report=bug_report,
+                smt_timeout=smt_timeout,
+                smt_retry_limit=smt_retry_limit,
+            )
+
+        class Task:
+            ...
+
+        class CreateAndRunTask(Task):
+            _test: str
+            _id: str | None
+
+            def __init__(self, test: str, id: str | None):
+                self._test = test
+                self._id = id
+
+        class RunTask(Task):
+            _proof: Proof
+            _port: int
+
+            def __init__(self, proof: Proof, port: int):
+                self._proof = proof
+                self._port = port
+
+        results = []
+        futures: dict[str, Future] = {}
+        executor = ThreadPoolExecutor(max_workers=workers)
+
+        tasks: list[Task] = [CreateAndRunTask(test, id) for test, id in tests]
+        base_proofs = {test: test for test, _ in tests}
+        servers = {}
+
+        for test, _ in tests:
+            servers[test] = create_server()
+
+        while len(tasks) > 0 or len(futures) > 0:
+            if len(tasks) > 0:
+                task = tasks.pop()
 
-        apr_proofs = dict(zip(tests, _apr_proofs, strict=True))
-        return apr_proofs
+                if type(task) is CreateAndRunTask:
+                    contract, method = task._test.split('.')
+
+                    futures[task._test] = executor.submit(_init_and_run_proof, (contract, method, task._id))
+
+                elif type(task) is RunTask:
+                    futures[task._proof.id] = executor.submit(_run_proof, (task._port, task._proof))
```

Comment:
Consider something like

```suggestion
                futures[task.name] = task.submit(executor)
```

where `name` and `submit` are abstract members of `Task`.

Reply:
Done.